Document and change types for panel move() direction

gordlin opened the pull request:
### Related Item(s)
Issue #1664 

### Changes
- [DOCS] Change JSDoc for PanelAPI/PanelInstance `move()` functions to describe acceptable options for direction (`"left"`/`"right"`).
- [REFACTOR] Change type of `direction` parameter from `string` to `"left" | "right"`, to be clearer.
- [DOCS] Modify panels.md to indicate the above change in parameter type.

### Testing
Steps:
1. In the "samples" demo site, go to `02. Lots of panels`. 
2. Try all of the following: Move the leftmost panel all the way to the right; swap the rightmost two panels repeatedly, then the leftmost two; move the rightmost panel all the way to the left; close the leftmost panel, then reopen it.
4. All previously-working panel movement interactions should still work, as seen.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)
<!-- Reviewable:end -->

james-rae reviewed (changes requested):
> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @gordlin)

___
*[`src/api/panel.ts` line 307 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFB5S85u8g4N9ag8Sr:-NzFB5S85u8g4N9ag8Ss:b8lzuhs) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L307)):*
> ```XML
>     move(
>         value: string | PanelInstance,
>         direction: "left" | "right"
> ```

do you have Prettier set-up in your VSCode? I'm surprised you were able to save without it changing the quotes to single-quotes.
___
*[`src/api/panel.ts` line 316 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFCJK84j-VwSAi9eAE:-NzFCJK84j-VwSAi9eAF:b-nasycf) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L316)):*
> ```XML
>         }
> 
>         this.panelStore.movePanel(panel, direction);
> ```

the code itself looks good, but you could probably turn the hardcoded `left | right`s into a `type`. If you declare it in `panel-store.ts` and export it, you can just reference the type in the methods you've adjusted, and can also use it in `panelStore.movePanel()`.  This means if someone adjusts the type it will be reflected everywhere.


<!-- Sent from Reviewable.io -->

milespetrov reviewed (approved):
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: all files reviewed, 2 unresolved discussions (waiting on @gordlin)



<!-- Sent from Reviewable.io -->

gordlin commented:
> > Reviewed 1 of 3 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)_ status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @gordlin)
> 
> _[`src/api/panel.ts` line 307 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFB5S85u8g4N9ag8Sr:-NzFB5S85u8g4N9ag8Ss:b8lzuhs) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L307)):_
> 
> > ```
> >     move(
> >         value: string | PanelInstance,
> >         direction: "left" | "right"
> > ```
> 
> do you have Prettier set-up in your VSCode? I'm surprised you were able to save without it changing the quotes to single-quotes.
> 
> _[`src/api/panel.ts` line 316 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFCJK84j-VwSAi9eAE:-NzFCJK84j-VwSAi9eAF:b-nasycf) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L316)):_
> 
> > ```
> >         }
> > 
> >         this.panelStore.movePanel(panel, direction);
> > ```
> 
> the code itself looks good, but you could probably turn the hardcoded `left | right`s into a `type`. If you declare it in `panel-store.ts` and export it, you can just reference the type in the methods you've adjusted, and can also use it in `panelStore.movePanel()`. This means if someone adjusts the type it will be reflected everywhere.

I added the new type and refactored the code. 

I did have Prettier installed, but it looks like I needed to set it to 'auto-format on save' within VSCode. Shouldn't be an issue in the future.

szczz reviewed (approved):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: all files reviewed, 3 unresolved discussions (waiting on @gordlin and @james-rae)

___
*[`src/api/panel-instance.ts` line 302 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzUG6x88NtZDcAcSv9L:-NzUG6x88NtZDcAcSv9M:btf061q) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L302)):*
> ```XML
>      * This is a proxy to `InstanceAPI.panel.move(...)`.
>      *
>      * @param {PanelDirection} direction the direction of movement, either "left" or "right".
> ```

Crimes against grammar. :oncoming_police_car: 


<!-- Sent from Reviewable.io -->

james-rae reviewed (changes requested):
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: all files reviewed, 4 unresolved discussions (waiting on @gordlin and @szczz)

___
*[`docs/api-guides/panels.md` line 65 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzYdU5142HfyZxoSTB-:-NzYdU52DQL0Y4qs126U:b-pftj2f) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/docs/api-guides/panels.md#L65)):*
> ```Markdown
> * `close()` - closes the panel. This means that the panel is no longer in the stack, but its `PanelInstance` object is still accessible for future use.
> * `minimize()` - minimizes the panel.
> * `move(direction: PanelDirection)` - moves the panel by one position in the specified direction (left or right), if possible.
> ```

Given this is a raw markdown file and not JSDoc, I don't think `PanelDirection` will get hotlinked to any definition. So maybe here change back to `string` or `"left" | "right"`. Whichever you chose, make sure the `PanelAPI.move` line lower down this file is the same
___
*[`src/api/panel.ts` line 316 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFCJK84j-VwSAi9eAE:-NzYeRi54CED9XZ0AYhc:bbdm64g) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L316)):*
<details><summary><i>Previously, gordlin (Gordon Lin) wrote…</i></summary><blockquote>

I added the new type and refactored the code. 

I did have Prettier installed, but it looks like I needed to set it to 'auto-format on save' within VSCode. Shouldn't be an issue in the future.

</blockquote></details>

I think `movePanel()` in the panel store is still typed as `string`
___
*[`src/api/panel-instance.ts` line 17 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzYdw575YN2Fb9Qk_qg:-NzYdw575YN2Fb9Qk_qh:b5vb95m) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L17)):*
> ```XML
>     PanelConfigRoute,
>     PanelConfigScreens,
>     PanelDirection,
> ```

Crime against the alphabet.
___
*[`src/api/panel-instance.ts` line 302 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzUG6x88NtZDcAcSv9L:-NzYdz-m9h0OL4z5mJIV:bv7n01c) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L302)):*
<details><summary><i>Previously, szczz (Matt Szczerba) wrote…</i></summary><blockquote>

Crimes against grammar. :oncoming_police_car:
</blockquote></details>

I think this might be correct? Looks weird in reviewable but in VSCode the first `direction` would be coloured to match the var name, and the remaining text is the text.  The doc-page generated should respect the convention as well. 
Unless there is a grammar mistake I'm missing. 


<!-- Sent from Reviewable.io -->

james-rae commented:
*[`src/api/panel-instance.ts` line 17 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzYdw575YN2Fb9Qk_qg:-NzYh2Xw2nQqPA-V-KnI:bq48fga) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L17)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

Crime against the alphabet.
</blockquote></details>

I see there are older crimes here as well.  So we can put half blame on past students.


<!-- Sent from Reviewable.io -->

gordlin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: 2 of 5 files reviewed, 4 unresolved discussions (waiting on @james-rae, @milespetrov, and @szczz)

___
*[`docs/api-guides/panels.md` line 65 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzYdU5142HfyZxoSTB-:-NzZa7_I3fcQKjz9fPCe:bah2568) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/docs/api-guides/panels.md#L65)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

Given this is a raw markdown file and not JSDoc, I don't think `PanelDirection` will get hotlinked to any definition. So maybe here change back to `string` or `"left" | "right"`. Whichever you chose, make sure the `PanelAPI.move` line lower down this file is the same
</blockquote></details>

Donethanks
___
*[`src/api/panel.ts` line 316 at r1](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzFCJK84j-VwSAi9eAE:-NzZaJOfCvVFQzDx_FtG:ba2iotn) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/64fc8d95202234db290941f30d2324470da88b01/src/api/panel.ts#L316)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

I think `movePanel()` in the panel store is still typed as `string`
</blockquote></details>

move() was as well. Both fixed.
___
*[`src/api/panel-instance.ts` line 17 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzYdw575YN2Fb9Qk_qg:-NzZaNuc6ks2qICi9CCb:b-law22v) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L17)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

I see there are older crimes here as well.  So we can put half blame on past students.
</blockquote></details>

Reordered.
___
*[`src/api/panel-instance.ts` line 302 at r2](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207#-NzUG6x88NtZDcAcSv9L:-NzZaPdC48NMlN5GRwrt:b-9kj5pc) ([raw file](https://github.com/ramp4-pcar4/ramp4-pcar4/blob/1496baff1d26c5e1bd95912408852e2bfc2daa6b/src/api/panel-instance.ts#L302)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

I think this might be correct? Looks weird in reviewable but in VSCode the first `direction` would be coloured to match the var name, and the remaining text is the text.  The doc-page generated should respect the convention as well. 
Unless there is a grammar mistake I'm missing.
</blockquote></details>

Yeah, I think the first word is the name of the parameter, and the rest is the description. It does look weird without a separator.


<!-- Sent from Reviewable.io -->

james-rae reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2207)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @gordlin)



<!-- Sent from Reviewable.io -->
